src/pagination.rs

```diff
@@ -0,0 +1,106 @@
+use std::{fmt::Display, future::Future, ops::Deref};
+
+use async_graphql::{
+    connection::{self, Connection, ConnectionNameType, CursorType, EdgeNameType, EmptyFields},
+    ErrorExtensions, InputObject, OutputType,
+};
+use chrono::{DateTime, FixedOffset};
+
+#[derive(Default, InputObject)]
+pub struct ConnectionParams {
+    pub after: Option<String>,
+    pub before: Option<String>,
+    pub first: Option<i32>,
+    pub last: Option<i32>,
+}
+
+pub async fn cursor_pagination<
+    Cursor,
+    Node,
+    Query,
+    Result,
+    Edge,
+    EdgeCursor,
+    ConnectionName,
+    EdgeName,
+>(
+    params: Option<ConnectionParams>,
+    query: Query,
+    edge_cursor: EdgeCursor,
+) -> async_graphql::Result<
+    Connection<Cursor, Node, EmptyFields, EmptyFields, ConnectionName, EdgeName>,
+>
+where
+    Cursor: Clone + CursorType + Send + Sync,
+    <Cursor as CursorType>::Error: Display + Send + Sync + 'static,
+    Node: OutputType,
+    Query: FnOnce(Option<Cursor>, Option<Cursor>, Option<u64>) -> Result,
+    Result: Future<Output = std::result::Result<Vec<Node>, Edge>>,
+    Edge: Into<Box<dyn std::error::Error>> + Display,
+    EdgeCursor: Fn(&Node) -> Cursor,
+    ConnectionName: ConnectionNameType,
+    EdgeName: EdgeNameType,
+{
+    let params = params.unwrap_or_default();
+    connection::query(
+        params.after,
+        params.before,
+        params.first,
+        params.last,
+        |after, before, first, last| async move {
+            if before.is_some() && last.is_some() {
+                panic!("Cannot query 'before' with 'last'");
```

Comment:
We can't panic if the client specifies both before and last. We should show the user an error indicating that the query is invalid.

Reply:
Fixed in https://github.com/eigerco/quasar/pull/54/commits/54d371dd2b8e4e927527002f9600928c0638eef3